core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -0,0 +1,613 @@
+    ++val;
+#else
+    float tmp = __half2float(val);
+    ++tmp;
+    val       = __float2half(tmp);
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  half_t& operator--() {
+#ifdef __CUDA_ARCH__
+    --val;
+#else
+    float tmp = __half2float(val);
+    --tmp;
+    val     = __float2half(tmp);
+#endif
+    return *this;
+  }
+
+  // Postfix operators
+  KOKKOS_FUNCTION
+  half_t operator++(int) {
+    half_t tmp = *this;
+    operator++();
+    return tmp;
+  }
+
+  KOKKOS_FUNCTION
+  half_t operator--(int) {
+    half_t tmp = *this;
+    operator--();
+    return tmp;
+  }
+
+  // Binary operators
+  KOKKOS_FUNCTION
+  half_t& operator=(impl_type rhs) {
+    val = rhs;
+    return *this;
+  }
+
+  template <class T>
+  KOKKOS_FUNCTION half_t& operator=(T rhs) {
+    val = cast_to_half(rhs).val;
+    return *this;
+  }
+
+  // Compound operators
+  KOKKOS_FUNCTION
+  half_t& operator+=(half_t rhs) {
+#ifdef __CUDA_ARCH__
+    val += rhs.val;
+#else
+    val     = __float2half(__half2float(val) + __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  half_t& operator-=(half_t rhs) {
+#ifdef __CUDA_ARCH__
+    val -= rhs.val;
+#else
+    val     = __float2half(__half2float(val) - __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  half_t& operator*=(half_t rhs) {
+#ifdef __CUDA_ARCH__
+    val *= rhs.val;
+#else
+    val     = __float2half(__half2float(val) * __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  KOKKOS_FUNCTION
+  half_t& operator/=(half_t rhs) {
+#ifdef __CUDA_ARCH__
+    val /= rhs.val;
+#else
+    val     = __float2half(__half2float(val) / __half2float(rhs.val));
+#endif
+    return *this;
+  }
+
+  // Binary Arithmetic
+  KOKKOS_FUNCTION
+  half_t operator+(half_t rhs) const {
+    half_t tmp = *this;
+#ifdef __CUDA_ARCH__
+    tmp.val += rhs.val;
+#else
+    tmp.val = __float2half(__half2float(tmp.val) + __half2float(rhs.val));
+#endif
+    return tmp;
+  }
+
+  KOKKOS_FUNCTION
+  half_t operator-(half_t rhs) const {
+    half_t tmp = *this;
+#ifdef __CUDA_ARCH__
+    tmp.val -= rhs.val;
+#else
+    tmp.val = __float2half(__half2float(tmp.val) - __half2float(rhs.val));
+#endif
+    return tmp;
+  }
+
+  KOKKOS_FUNCTION
+  half_t operator*(half_t rhs) const {
+    half_t tmp = *this;
+#ifdef __CUDA_ARCH__
+    tmp.val *= rhs.val;
+#else
+    tmp.val = __float2half(__half2float(tmp.val) * __half2float(rhs.val));
+#endif
+    return tmp;
+  }
+
+  KOKKOS_FUNCTION
+  half_t operator/(half_t rhs) const {
+    half_t tmp = *this;
+#ifdef __CUDA_ARCH__
+    tmp.val /= rhs.val;
+#else
+    tmp.val = __float2half(__half2float(tmp.val) / __half2float(rhs.val));
+#endif
+    return tmp;
+  }
+
+  // Logical operators
+  KOKKOS_FUNCTION
+  bool operator!() const {
+#ifdef __CUDA_ARCH__
+    return !val;
+#else
+    return !__half2float(val);
+#endif
+  }
+
+  // NOTE: Loses short-circuit evaluation
+  KOKKOS_FUNCTION
+  bool operator&&(half_t rhs) const {
```

Comment:
Thank you, @nilber. The unit [tests](https://github.com/kokkos/kokkos/pull/3439/files#diff-6980b46765a569d9ae64dffdbe6a0db53b9ed49c8b9157236055a90add5b47afR261-R265) were written to check for simple short circuiting. I will update the unit tests with some more complex expressions that rely on operator precedence as well.

Reply:
what kind of bugs? We could collect feedback on this. But since we don't have implicit conversion to bool (only from) statements like a&&b wouldn't work otherwise and you would need to do bool(a)&&bool(b) instead. So I am in favor of this.

Reply:
In favor of keeping these logical operator overloads? Shall I update the unit tests with some more complex expressions involving these logical operators?

Reply:
I think I just merge it. Nevin did approve after all. And if its a real issue he shouldn't have :-)